// @cloudinary/url-gen/actions/adjust/ImproveAction.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../tslib.es6-18ed4f64.cjs');
var internal_Action = require('../../internal/Action.cjs');
var internal_qualifier_QualifierValue = require('../../internal/qualifier/QualifierValue.cjs');
var internal_qualifier_Qualifier = require('../../internal/qualifier/Qualifier.cjs');
require('../../qualifiers/flag/FlagQualifier.cjs');
require('../../internal/utils/dataStructureUtils.cjs');
require('../../internal/utils/unsupportedError.cjs');

/**
 * @description Defines how to improve an image by automatically adjusting image colors, contrast and brightness.</br>
 * <b>Learn more:</b> {@link https://cloudinary.com/documentation/image_transformations#image_improvement_effects | Image improvement effects}
 * @memberOf Actions.Adjust
 */
var ImproveAction = /** @class */ (function (_super) {
    tslib_es6.__extends(ImproveAction, _super);
    function ImproveAction() {
        return _super.call(this) || this;
    }
    /**
     *
     * @description The improve mode.
     * @param {Qualifiers.ImproveMode | string} value
     */
    ImproveAction.prototype.mode = function (value) {
        this.modeValue = value;
        return this;
    };
    /**
     * @description How much to blend the improved result with the original image, where 0 means only use the original and 100 means only use the improved result. (Range: 0 to 100, Server default: 100)
     * @param {number} value
     */
    ImproveAction.prototype.blend = function (value) {
        this.blendValue = value;
        return this;
    };
    ImproveAction.prototype.prepareQualifiers = function () {
        var qualifierValue = new internal_qualifier_QualifierValue.QualifierValue(['improve', this.modeValue, this.blendValue]).setDelimiter(':');
        this.addQualifier(new internal_qualifier_Qualifier.Qualifier('e', qualifierValue));
        return this;
    };
    return ImproveAction;
}(internal_Action.Action));

exports.ImproveAction = ImproveAction;
