// @cloudinary/url-gen/qualifiers/source/BaseSource.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * @memberOf Qualifiers.Source
 * @description An abstract class extended by all Source objects
 */
var BaseSource = /** @class */ (function () {
    function BaseSource() {
    }
    /**
     * @description Utility function to encode an asset publicID in an overlay
     * @protected
     * @example
     * encodeAssetPublicID('foo/bar'); // -> foo:bar
     */
    BaseSource.prototype.encodeAssetPublicID = function (publicID) {
        return publicID.replace(/\//g, ':');
    };
    /**
     * @description
     * Apply a transformation on the image source of the layer
     * @param {SDK.ImageTransformation} t An image transformation to apply to the layer
     * @returns {this}
     */
    BaseSource.prototype.transformation = function (t) {
        this._transformation = t;
        return this;
    };
    /**
     * @description Returns the Transformation of the source
     * @return {SDK.Transformation}
     */
    BaseSource.prototype.getTransformation = function () {
        return this._transformation;
    };
    return BaseSource;
}());

exports.BaseSource = BaseSource;
